### PR TITLE
Async write support for ORC

### PR DESCRIPTION
**Pull request by jihoonson**
Follow up to https://github.com/NVIDIA/spark-rapids/pull/11730.

This PR adds the async write support for the ORC format as well as an integration test for it. This PR is marked as a draft since it depends on https://github.com/NVIDIA/spark-rapids/pull/11855.

### CODE REVIEW

**Comment by jihoonson** [comment]
build

**Comment by jihoonson** [comment]
build

File: integration_tests/src/main/python/orc_write_test.py [hunk 1]
```diff
@@ -360,6 +360,23 @@ def create_empty_df(spark, path):
...
+hold_gpu_configs = [True, False]
+@pytest.mark.parametrize('hold_gpu', hold_gpu_configs, ids=idfn)
+def test_async_writer(spark_tmp_path, hold_gpu):
+    data_path = spark_tmp_path + '/ORC_DATA'
+    num_rows = 2048
+    num_cols = 10
+    orc_gen = [int_gen for _ in range(num_cols)]
+    gen_list = [('_c' + str(i), gen) for i, gen in enumerate(orc_gen)]
+    assert_gpu_and_cpu_writes_are_equal_collect(
+        lambda spark, path: gen_df(spark, gen_list, length=num_rows).coalesce(1).write.orc(path),
```
Comment (abellina) [inline integration_tests/src/main/python/orc_write_test.py:372]: why `coalesce(1)`?

**Comment by abellina** [comment]
Any indication that this is a performance improvement for Orc?

**Review by revans2 — COMMENTED** [review]
Looks good to me, but want to hear about the coalesce question from @abellina

File: tools/generated_files/351/supportedExprs.csv [hunk 2]
```diff
@@ -637,6 +637,12 @@ TransformKeys,S,`transform_keys`,None,project,result,NA,NA,NA,NA,NA,NA,NA,NA,NA,
 TransformValues,S,`transform_values`,None,project,argument,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,PS,NA,NA,NA,NA
 TransformValues,S,`transform_values`,None,project,function,S,S,S,S,S,S,S,S,PS,S,S,S,NS,NS,PS,PS,PS,NS,NS,NS
 TransformValues,S,`transform_values`,None,project,result,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,PS,NA,NA,NA,NA
+TruncDate,S,`trunc`,None,project,date,NA,NA,NA,NA,NA,NA,NA,S,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA
+TruncDate,S,`trunc`,None,project,format,NA,NA,NA,NA,NA,NA,NA,NA,NA,S,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA
+TruncDate,S,`trunc`,None,project,result,NA,NA,NA,NA,NA,NA,NA,S,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA
+TruncTimestamp,S,`date_trunc`,None,project,format,NA,NA,NA,NA,NA,NA,NA,NA,NA,S,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA
+TruncTimestamp,S,`date_trunc`,None,project,date,NA,NA,NA,NA,NA,NA,NA,NA,PS,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA
+TruncTimestamp,S,`date_trunc`,None,project,result,NA,NA,NA,NA,NA,NA,NA,NA,PS,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA
```
Comment (mythrocks) [inline tools/generated_files/351/supportedExprs.csv:645]: Please pardon the nitpick, but please consider removing the changes to `operatorsScore.csv` and `supportedExprs.csv`.   That part isn't related to your current change.  

I'll file a separate PR to fix this.

Comment (mythrocks) [reply tools/generated_files/351/supportedExprs.csv:645]: https://github.com/NVIDIA/spark-rapids/pull/11890)

File: integration_tests/src/main/python/orc_write_test.py (hunk 1 above)
Comment (mythrocks) [reply integration_tests/src/main/python/orc_write_test.py:372]: Sharp eye. 

My guess is that `coalesce(1)` isn't strictly required for the test to pass.

**Review by mythrocks — COMMENTED** [review]
LGTM, barring the accidental inclusion of `operatorsScore.csv` and `supportedExprs.csv`.

File: integration_tests/src/main/python/orc_write_test.py (hunk 1 above)
Comment (jihoonson) [reply integration_tests/src/main/python/orc_write_test.py]: Oh good catch. I think I just copy-pasted without looking. I copy-pasted this from `test_async_writer()` in `parquet_write_test.py`, which in turn was copied from another test in the same file. I fixed both async writer tests.

File: tools/generated_files/351/supportedExprs.csv (hunk 2 above)
Comment (jihoonson) [reply tools/generated_files/351/supportedExprs.csv]: Thanks for making a fix. Reverted unrelated change.

**Comment by jihoonson** [comment]
> Any indication that this is a performance improvement for Orc? I assume so, but asking if you have data.

I haven't, but agree it will be useful. Unfortunately, I'm quite stuck in other work right now. Do you mind if I run some benchmark later and post the results here?

**Comment by jihoonson** [comment]
build

**Comment by jihoonson** [comment]
build

**Comment by jihoonson** [comment]
The test failure is filed in https://github.com/NVIDIA/spark-rapids/issues/11897.

**Comment by jihoonson** [comment]
build

**Review by abellina — APPROVED** [review]

**Review by revans2 — APPROVED** [review]

File: integration_tests/src/main/python/orc_write_test.py [hunk 3]
```diff
@@ -360,6 +360,23 @@ def create_empty_df(spark, path):
...
+@pytest.mark.parametrize('hold_gpu', hold_gpu_configs, ids=idfn)
+def test_async_writer(spark_tmp_path, hold_gpu):
+    data_path = spark_tmp_path + '/ORC_DATA'
+    num_rows = 2048
+    num_cols = 10
+    orc_gen = [int_gen for _ in range(num_cols)]
+    gen_list = [('_c' + str(i), gen) for i, gen in enumerate(orc_gen)]
+    assert_gpu_and_cpu_writes_are_equal_collect(
+        lambda spark, path: gen_df(spark, gen_list, length=num_rows).write.orc(path),
+        lambda spark, path: spark.read.orc(path).orderBy([('_c' + str(i)) for i in range(num_cols)]),
```
[the remark this answers is not in the document]
Comment (gerashegalov) [reply integration_tests/src/main/python/orc_write_test.py:373]: nit: we have @ignore_order marker for this

[the remark this answers is not in the document]
Comment (jihoonson) [reply integration_tests/src/main/python/orc_write_test.py:373]: Oh, good to know. I will use this instead next time.

**Comment by jihoonson** [comment]
Thanks all for the review!

**Pull request merged**